assembly/src/lib.rs

```diff
@@ -68,39 +68,30 @@ const MAX_PROC_NAME_LEN: u8 = 100;
 /// input is provided to `push` masm operation without period separators.
 const HEX_CHUNK_SIZE: usize = 16;
 
-// MODULE PROVIDER
-// ================================================================================================
-
-/// The module provider is now a simplified version of a module cache. It is expected to evolve to
-/// a general solution for the module lookup.
-pub trait ModuleProvider {
-    /// Fetch a module AST from its ID
-    fn get_module(&self, id: &ProcedureId) -> Option<NamedModuleAst<'_>>;
-}
-
-// A default provider that won't resolve modules
-impl ModuleProvider for () {
-    fn get_module(&self, _id: &ProcedureId) -> Option<NamedModuleAst<'_>> {
-        None
-    }
-}
-
 // LIBRARY
 // ================================================================================================
 
-/// TODO: add docs
-pub trait Library {
-    type Module;
-
+/// A library definition that provides AST modules for the compilation process.
+///
+/// Its `IntoIterator` implementation will be used to provide the modules to the assembler.
+pub trait Library: IntoIterator<Item = Module> {
     /// Returns the root namespace of this library.
-    fn root_ns(&self) -> &str;
+    fn root_ns(&self) -> &LibraryNamespace;
 
     /// Returns the version number of this library.
+    // TODO should have a SEMVER well-formed struct instead of raw string.
     fn version(&self) -> &str;
 
-    /// Returns the module located at the specified path.
+    /// Append a pre-processed module to the library.
     ///
-    /// # Errors
-    /// Returns an error if the modules for the specified path does not exist in this library.
-    fn get_module(&self, module_path: &str) -> Result<&Self::Module, LibraryError>;
+    /// This function should be called only if module is guaranteed to have a valid namespace.
+    fn add_module_with_absolute_path(&mut self, module: Module);
+
+    /// Add a module AST with its path to the library.
+    fn add_module(&mut self, path: &ModulePath, ast: ModuleAst) {
+        self.add_module_with_absolute_path(Module {
+            path: path.to_absolute(self.root_ns()),
+            ast,
+        })
+    }
```

Comment:
I think this works, but I am curious what are the benefits of this over the approach outlined in the issue. I think compared to this approach, the other approach is more flexible for two reasons:

1. It does not assume how a `Library` is constructed - this is left fully to the implementers.
2. We do not need to consume the library to retrieve modules from it.

Neither of these are super important, but I do wonder what we gain here by giving up this flexibility.

Reply:
1. That is something to think about. The way I did was intended to be less flexible and restrict the users to create the library only the right way and use the provided `ModulePath` to guarantee that.
2. We will end up having to clone the `ModuleAst` to store it. It's a matter of preference; whenever I need to clone something, I prefer to take the type as owned in the signature so the users will know that will be reallocated (and sometimes they don't have use for the type after it is consumed).

Reply:
The second point makes sense! On the first point, I would have probably gone in a different direction, but I don't really have any good reasons for why we'd need this extra flexibility - so, let's keep it the way you have it.